### PR TITLE
system: esp32: mcuboot as zephyr app

### PR DESCRIPTION
**Pull request by marekmatej**
This WIP aims to replace the IDF bootloader with the MCUBoot together with sysbuild.

### CODE REVIEW

File: components/bootloader_support/src/bootloader_init.c [hunk 1]
```diff
@@ -90,6 +90,10 @@ void bootloader_enable_random(void)
 
 void bootloader_print_banner(void)
 {
-    ESP_LOGI(TAG, "ESP-IDF %s 2nd stage bootloader", IDF_VER);
-    ESP_LOGI(TAG, "compile time " __TIME__);
+#ifdef CONFIG_MCUBOOT
+	ESP_LOGI(TAG, "## MCUBoot %s ESP32 bootloader", IDF_VER);
```
Comment (sylvioalves) [inline components/bootloader_support/src/bootloader_init.c:94]: Would be possible to use `printk` or Zephyr's logging subsystem instead?

Comment (almir-okato) [reply components/bootloader_support/src/bootloader_init.c:94]: These kind of macros are mapped to MCUboot's logging functions. In my opinion we could keep as is.

Comment (gustavonihei) [inline components/bootloader_support/src/bootloader_init.c:94]:
```suggestion
	ESP_LOGI(TAG, "## MCUboot %s ESP32 bootloader", IDF_VER);
```
nit: Wrong stylization of **MCUboot**

Comment (marekmatej) [reply components/bootloader_support/src/bootloader_init.c:94]: In fact that code is not even used, removing it.

File: components/esp_rom/include/esp32/rom/uart.h [hunk 2]
```diff
@@ -122,7 +122,7 @@ typedef enum {
 } UartFlowCtrl;
 
 typedef enum {
-    EMPTY,
+    EMPTY_,
```
Comment (LucasTambor) [inline components/esp_rom/include/esp32/rom/uart.h:125]: Don't get why add this underline, and why it is at the end instead of the beginning like C3 and S2.

File: zephyr/esp32c3/CMakeLists.txt [hunk 3]
```diff
@@ -162,13 +162,42 @@ if(CONFIG_SOC_ESP32C3)
...
+      ../../components/bootloader_support/src/bootloader_clock_init.c
+      ../../components/bootloader_support/src/bootloader_console.c
+      ../../components/bootloader_support/src/bootloader_random.c
+      ../../components/bootloader_support/src/bootloader_random_esp32c3.c
+      ../../components/esp_hw_support/cpu_util.c
+      ../../components/esp_hw_support/port/esp32c3/cpu_util_esp32c3.c
+      ../../components/hal/mpu_hal.c
+      ../../components/esp_rom/patches/esp_rom_uart.c
+      ../../components/esp_rom/patches/esp_rom_sys.c
+      ../../components/hal/wdt_hal_iram.c
```
Comment (LucasTambor) [inline zephyr/esp32c3/CMakeLists.txt:186]: Shouldn't the wdt source be outside of the if(CONFIG_MCUBOOT)?

File: components/esp_rom/include/esp32/rom/uart.h (hunk 2 above)
Comment (marekmatej) [reply components/esp_rom/include/esp32/rom/uart.h:125]: This is due to collision with other 'EMPTY', defined as `#define EMPTY`. Underline placement is a typo with no meaning.

**Review by LucasTambor — APPROVED** [review]

**Comment by dean-at-evos** [comment]
TLDR; When this PR is implemented will I still have to maintain a seperate MCUboot build step in production workflows for ESP32-C3? When might this PR be implemented?

I have a question about this PR. I am currently working on a production/CI workflow for 3 products we currently have in development based on the ESP32-C3.
At the moment we build the MCUboot bootloader outside zephyr using the espressif fork of MCUboot (as documented here https://github.com/mcu-tools/mcuboot/blob/main/docs/readme-espressif.md)
Then flash the bootloader to target.
Then zephyr product app is flashed onto target.
Will this PR once implemented allow building the MCUboot bootloader + application in the app tree?

**Comment by marekmatej** [comment]
> TLDR; When this PR is implemented will I still have to maintain a seperate MCUboot build step in production workflows for ESP32-C3? When might this PR be implemented?
> 
> I have a question about this PR. I am currently working on a production/CI workflow for 3 products we currently have in development based on the ESP32-C3. At the moment we build the MCUboot bootloader outside zephyr using the espressif fork of MCUboot (as documented here https://github.com/mcu-tools/mcuboot/blob/main/docs/readme-espressif.md) Then flash the bootloader to target. Then zephyr product app is flashed onto target. Will this PR once implemented allow building the MCUboot bootloader + application in the app tree?

Hi, this implementation is aiming to use `SYSBUILD`, which handles multi image build and flash. The MCUboot to be build can be found in `bootloader/mcuboot` folder of your Zephyrproject. Only difference in the build process would be adding the sysbuild option in the west call like this: `west build -b <board> --sysbuild <path/to/app>`. You can choose setting it permanently by using the `west config build.sysbuild True`.

File: zephyr/esp32c3/include/sdkconfig.h [hunk 4]
```diff
@@ -14,3 +14,4 @@
 #define BTDM_CTRL_HCI_MODE_VHCI 1
 
 #define CONFIG_ESPTOOLPY_FLASHMODE_DIO 1
+#define CONFIG_PARTITION_TABLE_OFFSET 0x8000
```
Comment (sylvioalves) [inline zephyr/esp32c3/include/sdkconfig.h:17]: Upstream version hardcodes this to 0x10000 to all SoCs... perhaps to support secure boot properly.  SHould we keep the same approach?
```c
#define CONFIG_PARTITION_TABLE_OFFSET 0x10000
```

File: zephyr/esp32/CMakeLists.txt [hunk 5]
```diff
@@ -194,12 +188,39 @@ if(CONFIG_SOC_ESP32 OR CONFIG_SOC_ESP32_NET)
...
+      ../../components/esp_hw_support/cpu_util.c
+      ../../components/esp_rom/patches/esp_rom_uart.c
+      ../../components/esp_rom/patches/esp_rom_sys.c
+      ../../components/hal/mpu_hal.c
+      )
+  else()
+
+    zephyr_sources(
+      ../../components/esp_hw_support/regi2c_ctrl.c
+    )
```
Comment (sylvioalves) [inline zephyr/esp32/CMakeLists.txt:216]: Would you fix code alignment here and on the same lines for others SoCs?

File: zephyr/esp32s3/CMakeLists.txt [hunk 6]
```diff
@@ -128,12 +128,41 @@ if(CONFIG_SOC_ESP32S3)
...
+    )
+  endif()
+
 	zephyr_sources(
     ../../components/soc/esp32s3/gpio_periph.c
     ../../components/soc/esp32s3/rtc_io_periph.c
     ../../components/spi_flash/spi_flash_timing_tuning.c
     ../../components/spi_flash/esp32s3/spi_timing_config.c
-    ../../components/esp_hw_support/regi2c_ctrl.c
+	#   ../../components/esp_hw_support/regi2c_ctrl.c
```
Comment (sylvioalves) [inline zephyr/esp32s3/CMakeLists.txt:165]: Needed or not?

File: zephyr/esp32/include/sdkconfig.h [hunk 7]
```diff
@@ -9,6 +22,8 @@
 #define BTDM_CTRL_HCI_MODE_VHCI 1
 #define CONFIG_SPIRAM_ALLOW_BSS_SEG_EXTERNAL_MEMORY 1
 #define CONFIG_FREERTOS_UNICORE 1
-#define CONFIG_PARTITION_TABLE_OFFSET 0x8000
 #define CONFIG_IDF_FIRMWARE_CHIP_ID 0x0000
 #define CONFIG_SPIRAM_SIZE -1
+
+#define CONFIG_ESP32_XTAL_FREQ_40 1
```
Comment (LucasTambor) [inline zephyr/esp32/include/sdkconfig.h:28]: Why is this value different from the clock_frequency on cpu node in esp32.dts?

File: components/bootloader_support/src/bootloader_flash.c [hunk 8]
```diff
@@ -189,7 +189,7 @@ const void *bootloader_mmap(uint32_t src_addr, uint32_t size)
         return NULL; /* can't map twice */
     }
     if (size > MMU_SIZE) {
-        ESP_LOGE(TAG, "bootloader_mmap excess size %x", size);
+        ESP_LOGE(TAG, "bootloader_mmap excess size (Bytes) %d", size);
```
Comment (sylvioalves) [inline components/bootloader_support/src/bootloader_flash.c:192]: Perhaps `%d bytes`? Not mandatory.

File: components/log/log_noos.c [hunk 9]
```diff
@@ -7,18 +7,10 @@
...
 #include <zephyr/kernel.h>
 #endif
-#else 
-#ifdef __ZEPHYR__
-#include <zephyr/kernel.h>
-#endif
-#include "hal/cpu_hal.h"  // for cpu_hal_get_cycle_count()
-#endif
-
+#include "hal/cpu_hal.h"
```
Comment (sylvioalves) [inline components/log/log_noos.c:13]: Have you tested this specific change (cpu_hal.h) placement for all SoCs? AFAIK, C3 required a different include sorting, which caused build warnings.

File: zephyr/esp32/include/sdkconfig.h [hunk 10]
```diff
@@ -1,3 +1,16 @@
+#ifdef CONFIG_MCUBOOT
+
+#define BOOTLOADER_BUILD 1
+#define NDEBUG 1
+
+#define CONFIG_BOOTLOADER_FLASH_XMC_SUPPORT 1
+#define CONFIG_BOOTLOADER_WDT_TIME_MS 9000
```
Comment (sylvioalves) [inline zephyr/esp32/include/sdkconfig.h:7]: I think both this CONFIGs should not be hardcoded. It can be added into kconfig on main branch.

File: zephyr/esp32/src/boot/bootloader_init.c [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        size = 16;
+        break;
+    default:
+        size = 2;
+    }
+    Cache_Read_Disable(0);
+    /* Set flash chip size */
+    esp_rom_spiflash_config_param(g_rom_flashchip.device_id,
+        size * 0x100000, 0x10000, 0x1000, 0x100, 0xffff);
+    /* TODO: set mode */
```
Comment (sylvioalves) [inline zephyr/esp32/src/boot/bootloader_init.c:134]: Is this your comment and still needs attention or original esp-idf's?

File: zephyr/esp32s3/include/sdkconfig.h [hunk 12]
```diff
@@ -9,6 +21,6 @@
 #define BTDM_CTRL_HCI_MODE_VHCI 1
 #define CONFIG_SPIRAM_ALLOW_BSS_SEG_EXTERNAL_MEMORY 1
 #define CONFIG_FREERTOS_UNICORE 1
-#define CONFIG_PARTITION_TABLE_OFFSET 0x8000
-#define CONFIG_IDF_FIRMWARE_CHIP_ID 0x0000
+#define CONFIG_IDF_FIRMWARE_CHIP_ID 0x009
+#define CONFIG_IDF_TARGET_ESP32S3 1
```
Comment (sylvioalves) [inline zephyr/esp32s3/include/sdkconfig.h:25]: This is already set by Zephyr's kconfig. Should not be necessary, right?

File: zephyr/esp32/include/sdkconfig.h (hunk 7 above)
Comment (marekmatej) [reply zephyr/esp32/include/sdkconfig.h:28]: forgot to clean it up, fixed

File: components/bootloader_support/src/bootloader_flash.c (hunk 8 above)
Comment (marekmatej) [reply components/bootloader_support/src/bootloader_flash.c:192]: fixed

File: components/log/log_noos.c (hunk 9 above)
Comment (marekmatej) [reply components/log/log_noos.c:13]: I've tested on all SOCs and all builds without warnings, so I did this simplification in includes.

File: zephyr/esp32/src/boot/bootloader_init.c [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        /* wait for regulator to become stable */
+        esp_rom_delay_us(10);
+    }
+}
+
+static void bootloader_reset_mmu(void)
+{
+    /* completely reset MMU in case serial bootloader was running */
+    Cache_Read_Disable(0);
+#if !CONFIG_FREERTOS_UNICORE
```
[the remark this answers is not in the document]
Comment (marekmatej) [reply zephyr/esp32/src/boot/bootloader_init.c:72]: couldn't decide between 2nd and 3rd option so I left as is

File: zephyr/esp32/src/boot/bootloader_init.c (hunk 11 above)
Comment (marekmatej) [reply zephyr/esp32/src/boot/bootloader_init.c:134]: this piece is from idf-5.x and was left as is

File: zephyr/esp32s3/include/sdkconfig.h [hunk 14]
```diff
@@ -9,6 +21,6 @@
 #define BTDM_CTRL_HCI_MODE_VHCI 1
 #define CONFIG_SPIRAM_ALLOW_BSS_SEG_EXTERNAL_MEMORY 1
 #define CONFIG_FREERTOS_UNICORE 1
-#define CONFIG_PARTITION_TABLE_OFFSET 0x8000
-#define CONFIG_IDF_FIRMWARE_CHIP_ID 0x0000
+#define CONFIG_IDF_FIRMWARE_CHIP_ID 0x009
```
[the remark this answers is not in the document]
Comment (marekmatej) [reply zephyr/esp32s3/include/sdkconfig.h:24]: yes, but I think it would lead to re-work to something like `soc/espressif` orso, which needs more time and separate PR

File: zephyr/esp32/src/boot/bootloader_init.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (sylvioalves) [reply zephyr/esp32/src/boot/bootloader_init.c:72]: @LucasTambor,. any opinion? Or we keep as is and handle it in another PR?

File: components/esp_hw_support/port/esp32c3/regi2c_ctrl.h [hunk 15]
```diff
@@ -42,7 +42,7 @@ uint8_t rom_i2c_readReg_Mask(uint8_t block, uint8_t host_id, uint8_t reg_add, ui
 void rom_i2c_writeReg(uint8_t block, uint8_t host_id, uint8_t reg_add, uint8_t data);
 void rom_i2c_writeReg_Mask(uint8_t block, uint8_t host_id, uint8_t reg_add, uint8_t msb, uint8_t lsb, uint8_t data);
 
-#ifdef BOOTLOADER_BUILD
+#if defined(BOOTLOADER_BUILD) && !defined(CONFIG_MCUBOOT)
```
Comment (sylvioalves) [inline components/esp_hw_support/port/esp32c3/regi2c_ctrl.h:45]: Is this necessary? Does it cause build warnings?

Comment (marekmatej) [reply components/esp_hw_support/port/esp32c3/regi2c_ctrl.h:45]: Exactly!

File: zephyr/esp32/src/boot/bootloader_init.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (LucasTambor) [reply zephyr/esp32/src/boot/bootloader_init.c:72]: I think we can handle hardcoded configs in a new PR

**Review by sylvioalves — APPROVED** [review]
